### DISCUSSION TITLE
Issues Dynamically Creating Job

### DISCUSSION THREAD

**Discussion by mccow002** [quartznet/quartznet#1101] [Q&A] [answered]
I'm running quartz.net in a windows service, and I'd like to be able to dynamically create jobs, but when I try to schedule a job anywhere outside the .AddQuartz() function, I get this error: "Problem instantiating class 'MyClass: Cannot instantiate type which has no empty constructor (Parameter 'MyClass')'."

I have quartz configured in my startup just like the documentation. I have a class that has ISchedulerFactory injected. That class has the following code:

```
var jobKey = new JobKey("test-once-job");
var triggerKey = new TriggerKey("test-once-trigger");

var job = JobBuilder.Create<TestOnceJob>()
    .WithIdentity(jobKey)
    .WithDescription("Test Once Job")
    .Build();

var trigger = TriggerBuilder.Create()
    .WithIdentity(triggerKey)
    .ForJob(jobKey)
    .StartAt(DateTime.UtcNow.AddSeconds(10))
    .Build();

var scheduler = await _schedulerFactory.GetScheduler();
await scheduler.ScheduleJob(job, trigger);
```

The TestOnceJob class has ILogger injected into it.

But like I said, this immediately throws ""Problem instantiating class 'MyClass: Cannot instantiate type which has no empty constructor (Parameter 'MyClass')'."

**Comment by lahma** [comment, marked as answer]
The problem here is that your dynamic job type isn't registered to DI. Quartz resolves the job from DI and if that fails it tries to use the default constructor, this is where you get the exception about not having default constructor. If you change the default configuration to:

```csharp
q.UseMicrosoftDependencyInjectionJobFactory(options =>
{
    options.AllowDefaultConstructor = false;
});
```

The problem becomes more apparent when exception is thrown about unresolvable job. So you would need some way to register the types into DI before trying to schedule them inside Quartz.

**Reply by mccow002** [reply]
That was the issue. Thank you!

**Reply by lahma** [reply]
An update to this thread. As of 3.3 you no longer need to register jobs to DI container id you don't want. They will be scoped instances with automatic constructor injection.
